tiiuae/vmd
# Implement basic CLI client

## Overview

Implement Nix flake build system and a CLI interface according to current openAPI.

```
Usage: vmd-client [OPTIONS] --hostname <HOSTNAME> --port <PORT> --cacert <CACERT> --key <KEY> --cert <CERT> --output <OUTPUT> <COMMAND>

Commands:
  list
  info
  action
  help    Print this message or the help of the given subcommand(s)

Options:
      --hostname <HOSTNAME>
  -p, --port <PORT>
      --cacert <CACERT>
      --key <KEY>
      --cert <CERT>
      --verbose
      --output <OUTPUT>      [possible values: json, yaml, text]
  -h, --help                 Print help
  -V, --version              Print version
  ```

## Requirements

- [x] Root flake.nix with separate outputs for client and server
- [x] Client CLI
- [x] JSON output
- [x] YAML output
- [x] TEXT output

## Testing (NixOS)

1. Open development shell

1. Create certificates

```bash
make -C test/auth
```

2. Start server

command:

```bash
nix run .#packages.x86_64-linux.vmd-server -- \
    --hostname localhost \
    --port 8080 \
    --cacert ./test/auth/certs/sample-ca-crt.pem \
    --cert ./test/auth/certs/sample-vmd-server-crt.pem \
    --key ./test/auth/certs/sample-vmd-server-key.pem
```

expected output:

```shell
⚡    VMD Server   ⚡

📌 api-version:           1.0.1
📡 address:               https://localhost:8080
🚩 base-path:             /api/v1
🔑 private-key:           ./test/auth/certs/sample-vmd-server-key.pem
🔐 certificate:           ./test/auth/certs/sample-vmd-server-crt.pem
🔐 certificate-authority: ./test/auth/certs/sample-ca-crt.pem
🔐 ocsp:                  🚫
```


3. Run client

command:

```bash
nix run .#packages.x86_64-linux.vmd-client -- \
    --hostname localhost \
    --port 8080 \
    --cacert test/auth/certs/sample-ca-crt.pem \
    --cert test/auth/certs/sample-vmd-client-chain.pem \
    --key test/auth/certs/sample-vmd-client-key.pem \
    --output yaml \
    list
```

expected output:

```shell
- 1
- 2
- 3
```
